File: engines/csharp/src/Engine/Ids.cs
using System;
using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;
using System.Text.RegularExpressions;

namespace G2048.Engine;

/// <summary>Seeds, ULIDs and replay codes (SPEC §9).</summary>
public static partial class Ids
{
    public const string ReplayAlphabet = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789";
    private const string Crockford = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";

    /// <summary>uint32 seed from a CSPRNG.</summary>
    public static uint RandomSeed() => BinaryPrimitives.ReadUInt32LittleEndian(RandomNumberGenerator.GetBytes(4));

    public static string Ulid(DateTimeOffset now)
    {
        ulong t = (ulong)now.ToUnixTimeMilliseconds();
        Span<char> o = stackalloc char[26];
        for (int i = 9; i >= 0; i--)
        {
            o[i] = Crockford[(int)(t % 32)];
            t /= 32;
        }
        byte[] rnd = RandomNumberGenerator.GetBytes(16);
        for (int i = 0; i < 16; i++) o[10 + i] = Crockford[rnd[i] & 31];
        return new string(o);
    }

    [GeneratedRegex("^[0-9A-HJKMNP-TV-Z]{26}$")]
    private static partial Regex UlidRegex();

    public static bool IsUlid(string s) => UlidRegex().IsMatch(s);

    public static string GenerateReplayCode()
    {
        byte[] rnd = RandomNumberGenerator.GetBytes(12);
        var s = new char[12];
        for (int i = 0; i < 12; i++) s[i] = ReplayAlphabet[rnd[i] & 31];
        return FormatReplayCode(new string(s));
    }

    /// <summary>Upper-cases, strips non [A-Z0-9], and validates; null if invalid.</summary>
    public static string? NormalizeReplayCode(string input)
    {
        var sb = new StringBuilder();
        foreach (char c0 in input)
        {
            char c = char.ToUpperInvariant(c0);
            if ((c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9')) sb.Append(c);
        }
        string s = sb.ToString();
        if (s.Length != 12) return null;
        foreach (char c in s) if (ReplayAlphabet.IndexOf(c) < 0) return null;
        return s;
    }

    /// <summary>XXXX-XXXX-XXXX.</summary>
    public static string FormatReplayCode(string n) => $"{n[..4]}-{n[4..8]}-{n[8..12]}";
}

File: engines/go/engine/ids.go
package engine

import (
	"crypto/rand"
	"encoding/binary"
	"regexp"
	"strings"
	"time"
)

// ReplayAlphabet is the replay-code alphabet (SPEC §9).
const ReplayAlphabet = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"

const crockford = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"

func randomBytes(n int) []byte {
	b := make([]byte, n)
	if _, err := rand.Read(b); err != nil {
		panic(err)
	}
	return b
}

// RandomSeed returns a uint32 seed from a CSPRNG.
func RandomSeed() uint32 { return binary.LittleEndian.Uint32(randomBytes(4)) }

// ULID returns a new ULID for the given time.
func ULID(now time.Time) string {
	t := uint64(now.UnixMilli())
	var out [26]byte
	for i := 9; i >= 0; i-- {
		out[i] = crockford[t%32]
		t /= 32
	}
	rnd := randomBytes(16)
	for i := 0; i < 16; i++ {
		out[10+i] = crockford[rnd[i]&31]
	}
	return string(out[:])
}

var ulidRe = regexp.MustCompile(`^[0-9A-HJKMNP-TV-Z]{26}$`)

// IsULID reports whether s is a ULID.
func IsULID(s string) bool { return ulidRe.MatchString(s) }

// GenerateReplayCode returns a new formatted replay code.
func GenerateReplayCode() string {
	rnd := randomBytes(12)
	var s [12]byte
	for i := range s {
		s[i] = ReplayAlphabet[rnd[i]&31]
	}
	return FormatReplayCode(string(s[:]))
}

// NormalizeReplayCode upper-cases, strips non [A-Z0-9], and validates.
func NormalizeReplayCode(input string) (string, bool) {
	var sb strings.Builder
	for _, c := range strings.ToUpper(input) {
		if (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') {
			sb.WriteRune(c)
		}
	}
	s := sb.String()
	if len(s) != 12 {
		return "", false
	}
	for i := 0; i < 12; i++ {
		if strings.IndexByte(ReplayAlphabet, s[i]) < 0 {
			return "", false
		}
	}
	return s, true
}

// FormatReplayCode inserts dashes: XXXX-XXXX-XXXX.
func FormatReplayCode(n string) string { return n[0:4] + "-" + n[4:8] + "-" + n[8:12] }
